{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow\n",
    "tf = tensorflow.compat.v1\n",
    "tf.disable_eager_execution()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_inputs = 3\n",
    "n_neurons = 5\n",
    "\n",
    "X0 = tf.placeholder(dtype=tf.float32, shape=[None, n_inputs])\n",
    "X1 = tf.placeholder(dtype=tf.float32, shape=[None, n_inputs])\n",
    "\n",
    "Wx = tf.Variable(tf.random_normal(shape=[n_inputs, n_neurons], dtype=tf.float32))\n",
    "Wy = tf.Variable(tf.random_normal(shape=[n_neurons, n_neurons], dtype=tf.float32))\n",
    "b = tf.Variable(tf.zeros([1, n_neurons], dtype=tf.float32))\n",
    "\n",
    "Y0 = tf.tanh(tf.matmul(X0, Wx) + b)\n",
    "Y1 = tf.tanh(tf.matmul(Y0, Wy) + tf.matmul(X1, Wx) + b)\n",
    "\n",
    "init = tf.global_variables_initializer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Mini-batch:  4 instances\n",
    "X0_batch = np.array([\n",
    "    [0, 1, 2], [3, 4, 5], [6, 7, 8], [9, 0, 1]\n",
    "])\n",
    "X1_batch = np.array([\n",
    "    [9, 8, 7], [0, 0, 0], [6, 5, 4], [3, 2, 1]\n",
    "])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "在之前定义tf.variable里的W时只是定义了这个变量值初始化的方式的，但并没有真正对其初始化，直到运行了tf.global_variables_initializer().run()之后才完成了变量的初始化。\n",
    "sess.run([Y0, Y1], feed_dict={X0: X0_batch, X1: X1_batch}) 的意思是执行[Y0, Y1]函数，执行时将其中的X0和X1替换为X0_batch, X1_batch。\n",
    "下面的函数只是执行了一步，还并没有开始训练"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with tf.Session() as sess:\n",
    "    init.run()\n",
    "    Y0_val, Y1_val = sess.run([Y0, Y1], feed_dict={X0: X0_batch, X1: X1_batch})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.9933106  -0.53062135  0.823383   -0.99261296 -0.20229085]\n",
      " [ 0.99190515 -0.99763733  0.89730775 -1.         -0.99029636]\n",
      " [ 0.9902062  -0.99999094  0.94128704 -1.         -0.99992836]\n",
      " [-1.          0.48571843 -1.         -0.99999917 -0.99999946]]\n",
      "[[-0.9885508  -1.          0.9484859  -1.         -0.99894595]\n",
      " [ 0.6136913  -0.22664668  0.99953645 -0.9186043   0.99823034]\n",
      " [-0.98015636 -0.9999363   0.99858886 -1.         -0.8355101 ]\n",
      " [-0.98742557 -0.5547205  -0.9501999  -0.99999505 -0.83876127]]\n"
     ]
    }
   ],
   "source": [
    "print(Y0_val)\n",
    "print(Y1_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "X0 = tf.placeholder(dtype=tf.float32, shape=[None, n_inputs])\n",
    "X1 = tf.placeholder(dtype=tf.float32, shape=[None, n_inputs])\n",
    "\n",
    "# basic_cell = tf.nn.rnn_cell.BasicRNNCell(num_units=n_neurons)\n",
    "basic_cell = tensorflow.keras.layers.SimpleRNNCell(units=n_neurons)\n",
    "output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=[X0, X1], dtype=tf.float32)\n",
    "\n",
    "Y0, Y1 = output_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "上面这种方法虽然可行，但显然当时间迭代长的时候是一件不那么容易的事情，所以，我们换一种做法。\n",
    "\n",
    "tf.transpose(X, perm=[1, 0, 2]) 是一个张量转置的函数，在这里就是将第一维和第二维转置，第三维保持不变。交换维度后，第一个时间迭代将称为第一个维度。unstack函数将抽取第一个维度的张良；stack函数合并所有输出张量为一个张量"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_steps = 2\n",
    "\n",
    "X = tf.placeholder(dtype=tf.float32, shape=[None, n_steps, n_inputs])\n",
    "X_seqs = tf.unstack(tf.transpose(X, perm=[1, 0, 2]))\n",
    "basic_cell = tensorflow.keras.layers.SimpleRNNCell(units=n_neurons)\n",
    "output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n",
    "\n",
    "outputs = tf.transpose(tf.stack(output_seqs), perm=[1, 0, 2])   # shape [None, n_steps, n_neurons]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_batch = np.array([\n",
    "    [[0, 1, 2], [9, 8, 7]],\n",
    "    [[3, 4, 5], [0, 0, 0]], \n",
    "    [[6, 7, 8], [6, 5, 4]], \n",
    "    [[9, 0, 1], [3, 2, 1]]\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "FailedPreconditionError",
     "evalue": "Graph execution error:\n\nDetected at node 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp' defined at (most recent call last):\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 196, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 86, in _run_code\n      exec(code, run_globals)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n      app.launch_new_instance()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\traitlets\\config\\application.py\", line 976, in launch_instance\n      app.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 712, in start\n      self.io_loop.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 215, in start\n      self.asyncio_loop.run_forever()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 600, in run_forever\n      self._run_once()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 1896, in _run_once\n      handle._run()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 510, in dispatch_queue\n      await self.process_one()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 499, in process_one\n      await dispatch(*args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 406, in dispatch_shell\n      await result\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 730, in execute_request\n      reply_content = await reply_content\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 383, in do_execute\n      res = shell.run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 528, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2881, in run_cell\n      result = self._run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2936, in _run_cell\n      return runner(coro)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3135, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3338, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3398, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_5108\\443957389.py\", line 6, in <cell line: 6>\n      output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\engine\\base_layer_v1.py\", line 763, in __call__\n      outputs = call_fn(cast_inputs, *args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\layers\\rnn\\simple_rnn.py\", line 194, in call\n      output = h + backend.dot(prev_output, self.recurrent_kernel)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 2223, in dot\n      out = tf.matmul(x, y)\nNode: 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp'\nDetected at node 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp' defined at (most recent call last):\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 196, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 86, in _run_code\n      exec(code, run_globals)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n      app.launch_new_instance()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\traitlets\\config\\application.py\", line 976, in launch_instance\n      app.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 712, in start\n      self.io_loop.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 215, in start\n      self.asyncio_loop.run_forever()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 600, in run_forever\n      self._run_once()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 1896, in _run_once\n      handle._run()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 510, in dispatch_queue\n      await self.process_one()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 499, in process_one\n      await dispatch(*args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 406, in dispatch_shell\n      await result\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 730, in execute_request\n      reply_content = await reply_content\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 383, in do_execute\n      res = shell.run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 528, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2881, in run_cell\n      result = self._run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2936, in _run_cell\n      return runner(coro)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3135, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3338, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3398, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_5108\\443957389.py\", line 6, in <cell line: 6>\n      output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\engine\\base_layer_v1.py\", line 763, in __call__\n      outputs = call_fn(cast_inputs, *args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\layers\\rnn\\simple_rnn.py\", line 194, in call\n      output = h + backend.dot(prev_output, self.recurrent_kernel)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 2223, in dot\n      out = tf.matmul(x, y)\nNode: 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp'\n2 root error(s) found.\n  (0) FAILED_PRECONDITION: Could not find variable rnn_2/simple_rnn_cell_1/recurrent_kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Resource localhost/rnn_2/simple_rnn_cell_1/recurrent_kernel/class tensorflow::Var does not exist.\n\t [[{{node rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp}}]]\n\t [[transpose_1/_3]]\n  (1) FAILED_PRECONDITION: Could not find variable rnn_2/simple_rnn_cell_1/recurrent_kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Resource localhost/rnn_2/simple_rnn_cell_1/recurrent_kernel/class tensorflow::Var does not exist.\n\t [[{{node rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp}}]]\n0 successful operations.\n0 derived errors ignored.\n\nOriginal stack trace for 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp':\n  File \"d:\\Python\\Python310\\lib\\runpy.py\", line 196, in _run_module_as_main\n    return _run_code(code, main_globals, None,\n  File \"d:\\Python\\Python310\\lib\\runpy.py\", line 86, in _run_code\n    exec(code, run_globals)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n    app.launch_new_instance()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\traitlets\\config\\application.py\", line 976, in launch_instance\n    app.start()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 712, in start\n    self.io_loop.start()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 215, in start\n    self.asyncio_loop.run_forever()\n  File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 600, in run_forever\n    self._run_once()\n  File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 1896, in _run_once\n    handle._run()\n  File \"d:\\Python\\Python310\\lib\\asyncio\\events.py\", line 80, in _run\n    self._context.run(self._callback, *self._args)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 510, in dispatch_queue\n    await self.process_one()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 499, in process_one\n    await dispatch(*args)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 406, in dispatch_shell\n    await result\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 730, in execute_request\n    reply_content = await reply_content\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 383, in do_execute\n    res = shell.run_cell(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 528, in run_cell\n    return super().run_cell(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2881, in run_cell\n    result = self._run_cell(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2936, in _run_cell\n    return runner(coro)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n    coro.send(None)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3135, in run_cell_async\n    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3338, in run_ast_nodes\n    if await self.run_code(code, result, async_=asy):\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3398, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_5108\\443957389.py\", line 6, in <cell line: 6>\n    output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\deprecation.py\", line 357, in new_func\n    return func(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150, in error_handler\n    return fn(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\dispatch.py\", line 1082, in op_dispatch_handler\n    return dispatch_target(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\rnn.py\", line 1480, in static_rnn\n    (output, state) = call_cell()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\rnn.py\", line 1467, in <lambda>\n    call_cell = lambda: cell(input_, state)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\engine\\base_layer_v1.py\", line 763, in __call__\n    outputs = call_fn(cast_inputs, *args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\layers\\rnn\\simple_rnn.py\", line 194, in call\n    output = h + backend.dot(prev_output, self.recurrent_kernel)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150, in error_handler\n    return fn(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\dispatch.py\", line 1082, in op_dispatch_handler\n    return dispatch_target(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 2223, in dot\n    out = tf.matmul(x, y)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150, in error_handler\n    return fn(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\dispatch.py\", line 1082, in op_dispatch_handler\n    return dispatch_target(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\math_ops.py\", line 3635, in matmul\n    b = ops.convert_to_tensor(b, dtype_hint=a.dtype.base_dtype, name=\"b\")\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\profiler\\trace.py\", line 183, in wrapped\n    return func(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 1623, in convert_to_tensor\n    ret = conversion_func(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 2077, in _dense_var_to_tensor\n    return var._dense_var_to_tensor(dtype=dtype, name=name, as_ref=as_ref)  # pylint: disable=protected-access\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1438, in _dense_var_to_tensor\n    return self.value()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 582, in value\n    return self._read_variable_op()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 691, in _read_variable_op\n    result = read_and_set_handle()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 681, in read_and_set_handle\n    result = gen_resource_variable_ops.read_variable_op(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\gen_resource_variable_ops.py\", line 492, in read_variable_op\n    _, _, _op, _outputs = _op_def_library._apply_op_helper(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\op_def_library.py\", line 797, in _apply_op_helper\n    op = g._create_op_internal(op_type_name, inputs, dtypes=None,\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 3754, in _create_op_internal\n    ret = Operation(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 2133, in __init__\n    self._traceback = tf_stack.extract_stack_for_node(self._c_op)\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:1377\u001b[0m, in \u001b[0;36mBaseSession._do_call\u001b[1;34m(self, fn, *args)\u001b[0m\n\u001b[0;32m   1376\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> 1377\u001b[0m   \u001b[39mreturn\u001b[39;00m fn(\u001b[39m*\u001b[39;49margs)\n\u001b[0;32m   1378\u001b[0m \u001b[39mexcept\u001b[39;00m errors\u001b[39m.\u001b[39mOpError \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:1360\u001b[0m, in \u001b[0;36mBaseSession._do_run.<locals>._run_fn\u001b[1;34m(feed_dict, fetch_list, target_list, options, run_metadata)\u001b[0m\n\u001b[0;32m   1359\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_extend_graph()\n\u001b[1;32m-> 1360\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_call_tf_sessionrun(options, feed_dict, fetch_list,\n\u001b[0;32m   1361\u001b[0m                                 target_list, run_metadata)\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:1453\u001b[0m, in \u001b[0;36mBaseSession._call_tf_sessionrun\u001b[1;34m(self, options, feed_dict, fetch_list, target_list, run_metadata)\u001b[0m\n\u001b[0;32m   1451\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_call_tf_sessionrun\u001b[39m(\u001b[39mself\u001b[39m, options, feed_dict, fetch_list, target_list,\n\u001b[0;32m   1452\u001b[0m                         run_metadata):\n\u001b[1;32m-> 1453\u001b[0m   \u001b[39mreturn\u001b[39;00m tf_session\u001b[39m.\u001b[39;49mTF_SessionRun_wrapper(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_session, options, feed_dict,\n\u001b[0;32m   1454\u001b[0m                                           fetch_list, target_list,\n\u001b[0;32m   1455\u001b[0m                                           run_metadata)\n",
      "\u001b[1;31mFailedPreconditionError\u001b[0m: 2 root error(s) found.\n  (0) FAILED_PRECONDITION: Could not find variable rnn_2/simple_rnn_cell_1/recurrent_kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Resource localhost/rnn_2/simple_rnn_cell_1/recurrent_kernel/class tensorflow::Var does not exist.\n\t [[{{node rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp}}]]\n\t [[transpose_1/_3]]\n  (1) FAILED_PRECONDITION: Could not find variable rnn_2/simple_rnn_cell_1/recurrent_kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Resource localhost/rnn_2/simple_rnn_cell_1/recurrent_kernel/class tensorflow::Var does not exist.\n\t [[{{node rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp}}]]\n0 successful operations.\n0 derived errors ignored.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[1;32me:\\TPGU\\SimpleRNN.ipynb Cell 11\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/e%3A/TPGU/SimpleRNN.ipynb#X14sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mwith\u001b[39;00m tf\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m sess:\n\u001b[0;32m      <a href='vscode-notebook-cell:/e%3A/TPGU/SimpleRNN.ipynb#X14sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m     init\u001b[39m.\u001b[39mrun()\n\u001b[1;32m----> <a href='vscode-notebook-cell:/e%3A/TPGU/SimpleRNN.ipynb#X14sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m     outputs_val \u001b[39m=\u001b[39m outputs\u001b[39m.\u001b[39;49meval(feed_dict\u001b[39m=\u001b[39;49m{X: X_batch})\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py:995\u001b[0m, in \u001b[0;36mTensor.eval\u001b[1;34m(self, feed_dict, session)\u001b[0m\n\u001b[0;32m    971\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39meval\u001b[39m(\u001b[39mself\u001b[39m, feed_dict\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, session\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m    972\u001b[0m   \u001b[39m\"\"\"Evaluates this tensor in a `Session`.\u001b[39;00m\n\u001b[0;32m    973\u001b[0m \n\u001b[0;32m    974\u001b[0m \u001b[39m  Note: If you are not using `compat.v1` libraries, you should not need this,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    993\u001b[0m \u001b[39m    A numpy array corresponding to the value of this tensor.\u001b[39;00m\n\u001b[0;32m    994\u001b[0m \u001b[39m  \"\"\"\u001b[39;00m\n\u001b[1;32m--> 995\u001b[0m   \u001b[39mreturn\u001b[39;00m _eval_using_default_session(\u001b[39mself\u001b[39;49m, feed_dict, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgraph, session)\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py:5747\u001b[0m, in \u001b[0;36m_eval_using_default_session\u001b[1;34m(tensors, feed_dict, graph, session)\u001b[0m\n\u001b[0;32m   5743\u001b[0m   \u001b[39mif\u001b[39;00m session\u001b[39m.\u001b[39mgraph \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m graph:\n\u001b[0;32m   5744\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mCannot use the given session to evaluate tensor: \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   5745\u001b[0m                      \u001b[39m\"\u001b[39m\u001b[39mthe tensor\u001b[39m\u001b[39m'\u001b[39m\u001b[39ms graph is different from the session\u001b[39m\u001b[39m'\u001b[39m\u001b[39ms \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   5746\u001b[0m                      \u001b[39m\"\u001b[39m\u001b[39mgraph.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m-> 5747\u001b[0m \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrun(tensors, feed_dict)\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:967\u001b[0m, in \u001b[0;36mBaseSession.run\u001b[1;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[0;32m    964\u001b[0m run_metadata_ptr \u001b[39m=\u001b[39m tf_session\u001b[39m.\u001b[39mTF_NewBuffer() \u001b[39mif\u001b[39;00m run_metadata \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m    966\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 967\u001b[0m   result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_run(\u001b[39mNone\u001b[39;49;00m, fetches, feed_dict, options_ptr,\n\u001b[0;32m    968\u001b[0m                      run_metadata_ptr)\n\u001b[0;32m    969\u001b[0m   \u001b[39mif\u001b[39;00m run_metadata:\n\u001b[0;32m    970\u001b[0m     proto_data \u001b[39m=\u001b[39m tf_session\u001b[39m.\u001b[39mTF_GetBuffer(run_metadata_ptr)\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:1190\u001b[0m, in \u001b[0;36mBaseSession._run\u001b[1;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[0;32m   1187\u001b[0m \u001b[39m# We only want to really perform the run if fetches or targets are provided,\u001b[39;00m\n\u001b[0;32m   1188\u001b[0m \u001b[39m# or if the call is a partial run that specifies feeds.\u001b[39;00m\n\u001b[0;32m   1189\u001b[0m \u001b[39mif\u001b[39;00m final_fetches \u001b[39mor\u001b[39;00m final_targets \u001b[39mor\u001b[39;00m (handle \u001b[39mand\u001b[39;00m feed_dict_tensor):\n\u001b[1;32m-> 1190\u001b[0m   results \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_do_run(handle, final_targets, final_fetches,\n\u001b[0;32m   1191\u001b[0m                          feed_dict_tensor, options, run_metadata)\n\u001b[0;32m   1192\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   1193\u001b[0m   results \u001b[39m=\u001b[39m []\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:1370\u001b[0m, in \u001b[0;36mBaseSession._do_run\u001b[1;34m(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)\u001b[0m\n\u001b[0;32m   1367\u001b[0m   \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_call_tf_sessionprun(handle, feed_dict, fetch_list)\n\u001b[0;32m   1369\u001b[0m \u001b[39mif\u001b[39;00m handle \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m-> 1370\u001b[0m   \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_do_call(_run_fn, feeds, fetches, targets, options,\n\u001b[0;32m   1371\u001b[0m                        run_metadata)\n\u001b[0;32m   1372\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   1373\u001b[0m   \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_do_call(_prun_fn, handle, feeds, fetches)\n",
      "File \u001b[1;32md:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\client\\session.py:1396\u001b[0m, in \u001b[0;36mBaseSession._do_call\u001b[1;34m(self, fn, *args)\u001b[0m\n\u001b[0;32m   1391\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39monly supports NHWC tensor format\u001b[39m\u001b[39m'\u001b[39m \u001b[39min\u001b[39;00m message:\n\u001b[0;32m   1392\u001b[0m   message \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m (\u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mA possible workaround: Try disabling Grappler optimizer\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m   1393\u001b[0m               \u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mby modifying the config for creating the session eg.\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m   1394\u001b[0m               \u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39msession_config.graph_options.rewrite_options.\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m   1395\u001b[0m               \u001b[39m'\u001b[39m\u001b[39mdisable_meta_optimizer = True\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m-> 1396\u001b[0m \u001b[39mraise\u001b[39;00m \u001b[39mtype\u001b[39m(e)(node_def, op, message)\n",
      "\u001b[1;31mFailedPreconditionError\u001b[0m: Graph execution error:\n\nDetected at node 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp' defined at (most recent call last):\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 196, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 86, in _run_code\n      exec(code, run_globals)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n      app.launch_new_instance()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\traitlets\\config\\application.py\", line 976, in launch_instance\n      app.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 712, in start\n      self.io_loop.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 215, in start\n      self.asyncio_loop.run_forever()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 600, in run_forever\n      self._run_once()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 1896, in _run_once\n      handle._run()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 510, in dispatch_queue\n      await self.process_one()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 499, in process_one\n      await dispatch(*args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 406, in dispatch_shell\n      await result\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 730, in execute_request\n      reply_content = await reply_content\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 383, in do_execute\n      res = shell.run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 528, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2881, in run_cell\n      result = self._run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2936, in _run_cell\n      return runner(coro)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3135, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3338, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3398, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_5108\\443957389.py\", line 6, in <cell line: 6>\n      output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\engine\\base_layer_v1.py\", line 763, in __call__\n      outputs = call_fn(cast_inputs, *args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\layers\\rnn\\simple_rnn.py\", line 194, in call\n      output = h + backend.dot(prev_output, self.recurrent_kernel)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 2223, in dot\n      out = tf.matmul(x, y)\nNode: 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp'\nDetected at node 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp' defined at (most recent call last):\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 196, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"d:\\Python\\Python310\\lib\\runpy.py\", line 86, in _run_code\n      exec(code, run_globals)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n      app.launch_new_instance()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\traitlets\\config\\application.py\", line 976, in launch_instance\n      app.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 712, in start\n      self.io_loop.start()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 215, in start\n      self.asyncio_loop.run_forever()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 600, in run_forever\n      self._run_once()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 1896, in _run_once\n      handle._run()\n    File \"d:\\Python\\Python310\\lib\\asyncio\\events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 510, in dispatch_queue\n      await self.process_one()\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 499, in process_one\n      await dispatch(*args)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 406, in dispatch_shell\n      await result\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 730, in execute_request\n      reply_content = await reply_content\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 383, in do_execute\n      res = shell.run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 528, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2881, in run_cell\n      result = self._run_cell(\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2936, in _run_cell\n      return runner(coro)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3135, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3338, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3398, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_5108\\443957389.py\", line 6, in <cell line: 6>\n      output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\engine\\base_layer_v1.py\", line 763, in __call__\n      outputs = call_fn(cast_inputs, *args, **kwargs)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\layers\\rnn\\simple_rnn.py\", line 194, in call\n      output = h + backend.dot(prev_output, self.recurrent_kernel)\n    File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 2223, in dot\n      out = tf.matmul(x, y)\nNode: 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp'\n2 root error(s) found.\n  (0) FAILED_PRECONDITION: Could not find variable rnn_2/simple_rnn_cell_1/recurrent_kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Resource localhost/rnn_2/simple_rnn_cell_1/recurrent_kernel/class tensorflow::Var does not exist.\n\t [[{{node rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp}}]]\n\t [[transpose_1/_3]]\n  (1) FAILED_PRECONDITION: Could not find variable rnn_2/simple_rnn_cell_1/recurrent_kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Resource localhost/rnn_2/simple_rnn_cell_1/recurrent_kernel/class tensorflow::Var does not exist.\n\t [[{{node rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp}}]]\n0 successful operations.\n0 derived errors ignored.\n\nOriginal stack trace for 'rnn_2/simple_rnn_cell_1/MatMul_1/ReadVariableOp':\n  File \"d:\\Python\\Python310\\lib\\runpy.py\", line 196, in _run_module_as_main\n    return _run_code(code, main_globals, None,\n  File \"d:\\Python\\Python310\\lib\\runpy.py\", line 86, in _run_code\n    exec(code, run_globals)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n    app.launch_new_instance()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\traitlets\\config\\application.py\", line 976, in launch_instance\n    app.start()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 712, in start\n    self.io_loop.start()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 215, in start\n    self.asyncio_loop.run_forever()\n  File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 600, in run_forever\n    self._run_once()\n  File \"d:\\Python\\Python310\\lib\\asyncio\\base_events.py\", line 1896, in _run_once\n    handle._run()\n  File \"d:\\Python\\Python310\\lib\\asyncio\\events.py\", line 80, in _run\n    self._context.run(self._callback, *self._args)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 510, in dispatch_queue\n    await self.process_one()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 499, in process_one\n    await dispatch(*args)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 406, in dispatch_shell\n    await result\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 730, in execute_request\n    reply_content = await reply_content\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 383, in do_execute\n    res = shell.run_cell(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 528, in run_cell\n    return super().run_cell(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2881, in run_cell\n    result = self._run_cell(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2936, in _run_cell\n    return runner(coro)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n    coro.send(None)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3135, in run_cell_async\n    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3338, in run_ast_nodes\n    if await self.run_code(code, result, async_=asy):\n  File \"d:\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3398, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_5108\\443957389.py\", line 6, in <cell line: 6>\n    output_seqs, states = tf.nn.static_rnn(cell=basic_cell, inputs=X_seqs, dtype=tf.float32)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\deprecation.py\", line 357, in new_func\n    return func(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150, in error_handler\n    return fn(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\dispatch.py\", line 1082, in op_dispatch_handler\n    return dispatch_target(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\rnn.py\", line 1480, in static_rnn\n    (output, state) = call_cell()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\rnn.py\", line 1467, in <lambda>\n    call_cell = lambda: cell(input_, state)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\engine\\base_layer_v1.py\", line 763, in __call__\n    outputs = call_fn(cast_inputs, *args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\layers\\rnn\\simple_rnn.py\", line 194, in call\n    output = h + backend.dot(prev_output, self.recurrent_kernel)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150, in error_handler\n    return fn(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\dispatch.py\", line 1082, in op_dispatch_handler\n    return dispatch_target(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 2223, in dot\n    out = tf.matmul(x, y)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150, in error_handler\n    return fn(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\util\\dispatch.py\", line 1082, in op_dispatch_handler\n    return dispatch_target(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\math_ops.py\", line 3635, in matmul\n    b = ops.convert_to_tensor(b, dtype_hint=a.dtype.base_dtype, name=\"b\")\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\profiler\\trace.py\", line 183, in wrapped\n    return func(*args, **kwargs)\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 1623, in convert_to_tensor\n    ret = conversion_func(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 2077, in _dense_var_to_tensor\n    return var._dense_var_to_tensor(dtype=dtype, name=name, as_ref=as_ref)  # pylint: disable=protected-access\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1438, in _dense_var_to_tensor\n    return self.value()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 582, in value\n    return self._read_variable_op()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 691, in _read_variable_op\n    result = read_and_set_handle()\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 681, in read_and_set_handle\n    result = gen_resource_variable_ops.read_variable_op(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\ops\\gen_resource_variable_ops.py\", line 492, in read_variable_op\n    _, _, _op, _outputs = _op_def_library._apply_op_helper(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\op_def_library.py\", line 797, in _apply_op_helper\n    op = g._create_op_internal(op_type_name, inputs, dtypes=None,\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 3754, in _create_op_internal\n    ret = Operation(\n  File \"d:\\Python\\Python310\\lib\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 2133, in __init__\n    self._traceback = tf_stack.extract_stack_for_node(self._c_op)\n"
     ]
    }
   ],
   "source": [
    "with tf.Session() as sess:\n",
    "    init.run()\n",
    "    outputs_val = outputs.eval(feed_dict={X: X_batch})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "301a17a29b57d3836b7901af1621afd6d2b1f2298b9c7949191147cf2fea93e3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
